torch_geometric/nn/to_hetero_module.py

```diff
@@ -9,42 +9,96 @@
 from torch_geometric.typing import EdgeType, NodeType, OptTensor
 from torch_geometric.utils import scatter
 
+from .dense import HeteroLinear, Linear
+
+
+class SubParam():
+    def __init__(self, param: torch.nn.Parameter, idx: int):
+        self.idx = idx
+        self.param = param
+
+    def get_data(self):
+        return self.param[self.idx].data
+
+    def set_data(self, data):
+        if self.param.data[self.idx].shape == data.t().shape:
+            self.param.data[self.idx] = data.t()
+        else:
+            self.param.data[self.idx] = data
+
+    data = property(get_data, set_data)
+
+
+class DummyLinear():
+    def __init__(self, get_type: Union[NodeType, EdgeType],
+                 types: Union[List[NodeType],
+                              List[EdgeType]], hetero_module: torch.nn.Module):
+        self.type = get_type
+        self.types = types
+        self.idx = types.index(get_type)
+        self.in_channels = hetero_module.in_channels
+        self.out_channels = hetero_module.out_channels
+        self.hetero_weight = hetero_module.weight
+        self.use_bias = hetero_module.kwargs.get('bias', True)
+
+        self.hetero_bias = hetero_module.bias
+
+    @property
+    def weight(self):
+        return SubParam(self.hetero_weight, self.idx)
+
+    @property
+    def bias(self):
+        if self.use_bias:
+            return SubParam(self.hetero_bias, self.idx)
+        else:
+            return None
+
 
 class ToHeteroLinear(torch.nn.Module):
     def __init__(
         self,
         module: torch.nn.Module,
         types: Union[List[NodeType], List[EdgeType]],
     ):
-        from torch_geometric.nn import HeteroLinear, Linear
-
         super().__init__()
 
         self.types = types
 
         if isinstance(module, Linear):
-            in_channels = module.in_channels
-            out_channels = module.out_channels
-            bias = module.bias is not None
+            self.in_channels = module.in_channels
+            self.out_channels = module.out_channels
 
         elif isinstance(module, torch.nn.Linear):
-            in_channels = module.in_features
-            out_channels = module.out_features
-            bias = module.bias is not None
+            self.in_channels = module.in_features
+            self.out_channels = module.out_features
 
         else:
             raise ValueError(f"Expected 'Linear' module (got '{type(module)}'")
 
-        # TODO: Need to handle `in_channels=-1` case.
-        # TODO We currently assume that `x` is sorted according to `type`.
+        self.bias = module.bias is not None
+        if self.in_channels == -1:
+            self.dim_dict = None
         self.hetero_module = HeteroLinear(
-            in_channels,
-            out_channels,
+            self.in_channels,
+            self.out_channels,
             num_types=len(types),
             is_sorted=True,
-            bias=bias,
+            bias=self.bias,
         )
 
+    def __getitem__(
```

Comment:
This seems more like a hack to move this to the current structure. Wouldn't it make more sense to not expect working with dictionaries now whenever we see a `Linear` layer?

Reply:
it is definitely a pythonic hack. took quite a bit of wrestling to get it working but my general belief is that my PR should accelerate to_hetero with zero code changes from the user needed or any changes in user experience and object behavior. I believe this kind of behavior you have set in the test is a valid thing to expect from a user POV and this python setup works.